tests/components/device_tracker/test_unifi_direct.py

```diff
@@ -0,0 +1,169 @@
+
+    def teardown_method(self, _):
+        """Stop everything that was started."""
+        self.hass.stop()
+        try:
+            os.remove(self.hass.config.path(device_tracker.YAML_DEVICES))
+        except FileNotFoundError:
+            pass
+
+    @mock.patch(scanner_path,
+                return_value=mock.MagicMock())
+    def test_get_scanner(self, unifi_mock):  \
+            # pylint: disable=invalid-name
+        """Test creating an Unifi direct scanner with a password."""
+        conf_dict = {
+            DOMAIN: {
+                CONF_PLATFORM: 'unifi_direct',
+                CONF_HOST: 'fake_host',
+                CONF_USERNAME: 'fake_user',
+                CONF_PASSWORD: 'fake_pass',
+                CONF_TRACK_NEW: True,
+                CONF_CONSIDER_HOME: timedelta(seconds=180)
+            }
+        }
+
+        with assert_setup_component(1, DOMAIN):
+            assert setup_component(self.hass, DOMAIN, conf_dict)
+
+        conf_dict[DOMAIN][CONF_PORT] = 22
+        self.assertEqual(unifi_mock.call_args, mock.call(conf_dict[DOMAIN]))
+
+    @patch('pexpect.pxssh.pxssh')
+    def test_get_device_name(self, mock_ssh):
+        """"Testing MAC matching."""
+        conf_dict = {
+            DOMAIN: {
+                CONF_PLATFORM: 'unifi_direct',
+                CONF_HOST: 'fake_host',
+                CONF_USERNAME: 'fake_user',
+                CONF_PASSWORD: 'fake_pass',
+                CONF_PORT: 22,
+                CONF_TRACK_NEW: True,
+                CONF_CONSIDER_HOME: timedelta(seconds=180)
+            }
+        }
+        mock_ssh.return_value.before = load_fixture('unifi_direct.txt')
+        scanner = get_scanner(self.hass, conf_dict)
+        devices = scanner.scan_devices()
+        self.assertEqual(23, len(devices))
+        self.assertEqual("iPhone",
+                         scanner.get_device_name("98:00:c6:56:34:12"))
+        self.assertEqual("iPhone",
+                         scanner.get_device_name("98:00:C6:56:34:12"))
+
+    @patch('pexpect.pxssh.pxssh.login')
+    def test_failed_to_log_in(self, mock_login):
+        """"Testing exception at login results in False."""
+        from pexpect import exceptions
+
+        conf_dict = {
+            DOMAIN: {
+                CONF_PLATFORM: 'unifi_direct',
+                CONF_HOST: 'fake_host',
+                CONF_USERNAME: 'fake_user',
+                CONF_PASSWORD: 'fake_pass',
+                CONF_PORT: 22,
+                CONF_TRACK_NEW: True,
+                CONF_CONSIDER_HOME: timedelta(seconds=180)
+            }
+        }
+
+        mock_login.side_effect = exceptions.EOF("Test")
+        scanner = get_scanner(self.hass, conf_dict)
+        self.assertFalse(scanner)
+
+    @patch('pexpect.pxssh.pxssh.sendline')
+    @patch('pexpect.pxssh.pxssh.prompt')
+    @patch('pexpect.pxssh.pxssh.login', autospec=True)
+    def test_to_get_update(self, mock_sendline, mock_prompt, mock_login):
```

Comment:
Same here.

Reply:
This is probably something else I don't understand about testing. Adding 
@patch('pexpect.pxssh.pxssh.logout') causes this test to fail. The exception isn't raised  and the scanner gets setup correctly. How is that?

Reply:
You have to be careful with the order of annotations of patches. It's done inside out, so that's why it failed I think. Put the logout patch at the top, and the corresponding mock argument is then the last argument.

Reply:
Looks like the order is incorrect for the other patches too.